Division of a BlockDataContainer of ImageData by a non-zero scalar gives NaN pixels

I believe this is related to Issue #798 but would like to flag it independently in case it's not.

If I define a `CIL.BlockDataContainer` made of `sirf.ImageData` blocks and divide it by a non-zero scalar, some (but not all) pixels are put to NaN.

```python
ImageData = am.domain_geometry()
ImageDatax3 = BlockGeometry(*[ImageData for _ in range(3) ] )
imagex3 = ImageDatax3.allocate('random')
sigma = 1.

print('Image has shape {}, that is {} pixels'.format(imagex3[0].as_array().shape, imagex3[0].as_array().size))

for image in imagex3 / sigma:

    print('New image has shape {}, that is {} pixels'.format(image.as_array().shape, image.as_array().size))
    print('Out of these, {} pixels are NaN'.format(np.isnan(image.as_array()).sum()))
```

outputs

```python
Image has shape (127, 127, 127), that is 2048383 pixels
New image has shape (127, 127, 127), that is 2048383 pixels
Out of these, 2722 pixels are NaN
New image has shape (127, 127, 127), that is 2048383 pixels
Out of these, 2722 pixels are NaN
New image has shape (127, 127, 127), that is 2048383 pixels
Out of these, 2722 pixels are NaN
```

Unfortunately I have a build error

```

/home/ofn77899/devel/buildVM/sources/SIRF/src/xSTIR/cSTIR/include/sirf/STIR/stir_data_containers.h:617:36: error: ‘class stir::ProjDataInMemory’ has no member named ‘end’
             while (iter != pd_ptr->end())
```

Any idea?

--- a/src/Registration/pReg/Reg.py
+++ b/src/Registration/pReg/Reg.py
@@ -520,6 +520,13 @@ class NiftiImageData(SIRF.ImageData):
         pyiutil.deleteDataHandle(h)
         return value
 
+    @property
+    def shape(self):
+        return self.dimensions()
+    @property
+    def dtype(self):
+        return numpy.float32
+
 
 class NiftiImageData3D(NiftiImageData):
     """Class for 3D nifti image data."""

--- a/src/common/SIRF.py
+++ b/src/common/SIRF.py
@@ -128,7 +128,7 @@ class DataContainer(ABC):
         out:   DataContainer to store the result to.
         '''
         if isinstance(other , ( Number, int, float, numpy.float32 )):
-            tmp = other + numpy.zeros(self.as_array().shape)
+            tmp = other + numpy.zeros(self.shape, self.dtype)
             other = self.copy()
             other.fill(tmp)
         assert_validities(self, other)
@@ -149,7 +149,7 @@ class DataContainer(ABC):
         out:   DataContainer to store the result to.
         '''
         if isinstance(other , ( Number, int, float, numpy.float32 )):
-            tmp = other + numpy.zeros_like(self.as_array().shape)
+            tmp = other + numpy.zeros(self.shape, self.dtype)
             other = self.copy()
             other.fill(tmp)
         assert_validities(self, other)
@@ -172,7 +172,7 @@ class DataContainer(ABC):
         out:   DataContainer to store the result to.
         '''
         if isinstance(other , ( Number, int, float, numpy.float32 )):
-            tmp = other + numpy.zeros(self.as_array().shape)
+            tmp = other + numpy.zeros(self.shape, self.dtype)
             other = self.copy()
             other.fill(tmp)
         assert_validities(self, other)
@@ -242,7 +242,7 @@ class DataContainer(ABC):
         other: DataContainer
         '''
         if isinstance(other , ( Number, int, float, numpy.float32 )):
-            tmp = other + numpy.zeros(self.as_array().shape)
+            tmp = other + numpy.zeros(self.shape, self.dtype)
             other = self.copy()
             other.fill(tmp)
         assert_validities(self, other)
@@ -539,6 +539,11 @@ class DataContainer(ABC):
         '''Returns the (total) size of the data array.'''
         return self.as_array().size
 
+    @property
+    def dtype(self):
+        '''return default type as float32'''
+        return numpy.float32
+    
 class ImageData(DataContainer):
     '''
     Image data ABC

--- a/src/xGadgetron/pGadgetron/Gadgetron.py
+++ b/src/xGadgetron/pGadgetron/Gadgetron.py
@@ -474,6 +474,8 @@ class ImageData(SIRF.ImageData):
         try_calling(pygadgetron.cGT_print_header(self.handle, im_num))
     @property
     def dtype(self):
+        if self.is_real():
+            return numpy.float32
         return numpy.complex64
     @property
     def shape(self):

--- a/src/xGadgetron/pGadgetron/Gadgetron.py
+++ b/src/xGadgetron/pGadgetron/Gadgetron.py
@@ -472,7 +472,10 @@ class ImageData(SIRF.ImageData):
     def print_header(self, im_num):
         """Print the header of one of the images. zero based."""
         try_calling(pygadgetron.cGT_print_header(self.handle, im_num))
-
+    @property
+    def dtype(self):
+        return numpy.float32
+        
 SIRF.ImageData.register(ImageData)
 
 
@@ -969,6 +972,12 @@ class AcquisitionData(DataContainer):
             tmp = value * numpy.ones(out.as_array().shape)
             out.fill(tmp)
         return out
+    @property
+    def shape(self):
+        return self.dimensions()
+    @property
+    def dtype(self):
+        return numpy.float32
     
     
 DataContainer.register(AcquisitionData)

--- a/src/xGadgetron/pGadgetron/Gadgetron.py
+++ b/src/xGadgetron/pGadgetron/Gadgetron.py
@@ -474,7 +474,10 @@ class ImageData(SIRF.ImageData):
         try_calling(pygadgetron.cGT_print_header(self.handle, im_num))
     @property
     def dtype(self):
-        return numpy.float32
+        return numpy.complex64
+    @property
+    def shape(self):
+        return self.dimensions()
         
 SIRF.ImageData.register(ImageData)
 
@@ -977,7 +980,7 @@ class AcquisitionData(DataContainer):
         return self.dimensions()
     @property
     def dtype(self):
-        return numpy.float32
+        return numpy.complex64
     
     
 DataContainer.register(AcquisitionData)

--- a/src/xSTIR/pSTIR/STIR.py
+++ b/src/xSTIR/pSTIR/STIR.py
@@ -560,6 +560,12 @@ class ImageData(SIRF.ImageData):
             moved_im.handle, proj_data.handle))
 
         return moved_im
+    @property
+    def shape(self):
+        return self.dimensions()
+    @property
+    def dtype(self):
+        return numpy.float32
 
 
 SIRF.ImageData.register(ImageData)
@@ -1007,6 +1013,13 @@ class AcquisitionData(DataContainer):
         info = pyiutil.charDataFromHandle(handle)
         pyiutil.deleteDataHandle(handle)
         return info
+
+    @property
+    def shape(self):
+        return self.dimensions()
+    @property
+    def dtype(self):
+        return numpy.float32
 
 
 DataContainer.register(AcquisitionData)

